### PR TITLE
many:  start completing snap.Info to match snapYaml expressivity, get rid of AppYaml usages

### PR DESCRIPTION
**Pull request by pedronis**
This is adding fields to snap.Info and snap.AppInfo to match the expressivity of snapYaml, and starting updating the places involved with AppYaml to use these instead (except old security that will go away).

This gives snappy.Snap a full snap.Info. We actually want snappy.Snap to go away but we really need to get rid of snapYaml first, in the mean time we need a snappy.Snap to offer both, done with a slightly ugly reparsing hack for now in some corner cases (mostly tests).

Still WIP atm:
* there's one failing test, need to implement snap.ValidateApp and port VerifyAppYaml tests
* service.go/services.go need a pass, add all used fields to AppInfo
* recheck if some XXXs can be dealt with immediately

### CODE REVIEW

File: snappy/security.go [hunk 1]
```diff
@@ -319,6 +320,13 @@ func findWhitespacePrefix(t string, s string) string {
 	return subs[1]
 }
 
+// XXX: what to use really?
+func getSecurityProfile2(s *snap.Info, appName, baseDir string) string {
```
Comment (pedronis) [inline snappy/security.go]: temp name, I think interfaces now has something for this

File: snap/info.go [hunk 2]
```diff
@@ -19,24 +19,31 @@
...
+	Type             Type
+	Architectures    []string
+	Channel          string
+	Description      string
+	Summary          string
+	LicenseAgreement string
+	LicenseVersion   string
+	Apps             map[string]*AppInfo
+	Plugs            map[string]*PlugInfo
+	Slots            map[string]*SlotInfo
```
Comment (niemeyer) [inline snap/info.go]: Can we please insert an empty line and then this comment before the following block:
```
// The information in these fields is not present inside the snap blob itself.
```
Then, let's please move Revision, Developer, and Channel to the top of this block.

File: snap/info.go [hunk 3]
```diff
@@ -65,8 +72,13 @@ type SlotInfo struct {
 type AppInfo struct {
 	Snap *Info
 
-	Name    string
-	Command string
-	Plugs   map[string]*PlugInfo
-	Slots   map[string]*SlotInfo
+	Name        string
+	Command     string
+	Description string
```
Comment (niemeyer) [inline snap/info.go]: Can we please take Description out from here until we make some better usage of it?  The current usage seems to be just for systemd services, which doesn't justify asking the user to type this in for every snap. The GUI is using the .desktop files.

File: snap/validate.go [hunk 4]
```diff
@@ -35,3 +35,13 @@ func ValidateName(name string) error {
 	}
 	return nil
 }
+
+// Validate verifies the content in the info.
+func Validate(info *Info) error {
```
Comment (niemeyer) [inline snap/validate.go]: How do you imagine using those? Should we validate when reading the info in at all times? If so, likely don't need those publicly?

File: snappy/binaries.go [hunk 5]
```diff
@@ -30,22 +30,33 @@ import (
 	"github.com/ubuntu-core/snappy/arch"
 	"github.com/ubuntu-core/snappy/dirs"
 	"github.com/ubuntu-core/snappy/osutil"
+	"github.com/ubuntu-core/snappy/snap"
 	"github.com/ubuntu-core/snappy/snap/snapenv"
 )
 
-// generate the name
-func generateBinaryName(m *snapYaml, app *AppYaml) string {
+func oldGenerateBinaryName(m *snapYaml, app *AppYaml) string {
```
Comment (niemeyer) [inline snappy/binaries.go]: Drop?

File: snappy/binaries.go [hunk 6]
```diff
@@ -30,22 +30,33 @@ import (
...
 	if app.Name == m.Name {
 		binName = filepath.Base(app.Name)
 	} else {
 		binName = fmt.Sprintf("%s.%s", m.Name, filepath.Base(app.Name))
 	}
+	return filepath.Join(dirs.SnapBinariesDir, binName)
+}
+
+// generate the name
+func generateBinaryName(app *snap.AppInfo) string {
```
Comment (niemeyer) [inline snappy/binaries.go]: Might become AppInfo.WrapperPath

File: snappy/binaries.go [hunk 7]
```diff
@@ -30,22 +30,33 @@ import (
...
+		binName = filepath.Base(app.Name)
+	} else {
+		binName = fmt.Sprintf("%s.%s", app.Snap.Name, filepath.Base(app.Name))
+	}
 
 	return filepath.Join(dirs.SnapBinariesDir, binName)
 }
 
-func binPathForBinary(pkgPath string, app *AppYaml) string {
+func binPathForBinary(pkgPath string, app *snap.AppInfo) string {
```
Comment (niemeyer) [inline snappy/binaries.go]: Might become AppInfo.CommandLine.

Note that the name is misleading. app.Command holds a command with arguments, not a path.

File: snappy/binaries.go [hunk 8]
```diff
@@ -128,28 +139,32 @@ ubuntu-core-launcher {{.UdevAppName}} {{.AaProfile}} {{.Target}} "$@"
...
 }
-func addPackageBinaries(m *snapYaml, baseDir string) error {
+
+func addPackageBinaries(s *snap.Info) error {
 	if err := os.MkdirAll(dirs.SnapBinariesDir, 0755); err != nil {
 		return err
 	}
 
-	for _, app := range m.Apps {
+	baseDir := BaseDir(s)
```
Comment (niemeyer) [inline snappy/binaries.go]: BaseDir is also a candidate for a generally useful Info.BaseDir method.

File: snappy/security.go [hunk 9]
```diff
@@ -319,6 +320,13 @@ func findWhitespacePrefix(t string, s string) string {
 	return subs[1]
 }
 
+// XXX: what to use really?
```
Comment (niemeyer) [inline snappy/security.go]: This is all going away imminently. Just tuck it away in the cheapest possible way, ideally grouping with related logic that will all being dropped with it. @zyga should be able to provide more details if you need it.

File: snappy/snap_local.go [hunk 10]
```diff
@@ -57,87 +59,115 @@ func NewInstalledSnap(yamlPath string) (*Snap, error) {
...
+	info.Channel = "stable"
+	info.Size = -1
+
+	if manifest != nil {
+		info.Revision = manifest.Revision
+		info.Developer = manifest.Developer
+		info.Channel = manifest.Channel
+		// store edits win!
+		info.Description = manifest.Description
+		// XXX: Summary doesn't exist in the store yet anyway
```
Comment (niemeyer) [inline snappy/snap_local.go]: It's coming any day now. If it's cheap let's please start adding it while driving by.

File: snappy/snap_local.go [hunk 11]
```diff
@@ -57,87 +59,115 @@ func NewInstalledSnap(yamlPath string) (*Snap, error) {
 	return snap, nil
 }
 
+func completeInfo(info *snap.Info, manifest *diskManifest) {
+	// XXX: we actually should not trust snap.yaml name!
+
+	info.Developer = SideloadedDeveloper
+	// default for compat with older installs
+	info.Channel = "stable"
+	info.Size = -1
```
Comment (niemeyer) [inline snappy/snap_local.go]: Zero seems like nice default here. It cannot possibly be valid.

File: snappy/snap_local.go [hunk 12]
```diff
@@ -57,87 +59,115 @@ func NewInstalledSnap(yamlPath string) (*Snap, error) {
 	return snap, nil
 }
 
+func completeInfo(info *snap.Info, manifest *diskManifest) {
+	// XXX: we actually should not trust snap.yaml name!
+
+	info.Developer = SideloadedDeveloper
```
Comment (niemeyer) [inline snappy/snap_local.go]: It doesn't have to be done now if it makes things more complex for you, but let's use an empty Developer when we don't know, instead of the magic marker.

File: snappy/snap_local.go [hunk 13]
```diff
@@ -57,87 +59,115 @@ func NewInstalledSnap(yamlPath string) (*Snap, error) {
...
-		snap.manifest = &manifest
 	}
 
-	return snap, nil
+	completeInfo(info, manifest)
+
+	// XXX: FIXME: just some tests need this atm
+	// override the package's idea of its version
+	// because that could have been rewritten on sideload
+	// and developer is empty sideloaded ones.
```
Comment (niemeyer) [inline snappy/snap_local.go]: I don't understand what this was supposed to do, but it feels like it doesn't make much sense in the modern world.

File: snappy/snap_local.go [hunk 14]
```diff
@@ -57,87 +59,115 @@ func NewInstalledSnap(yamlPath string) (*Snap, error) {
...
+	m.Version = filepath.Base(s.basedir)
```
Comment (niemeyer) [inline snappy/snap_local.go]: Irrelevant now, but awkward. We ask to generate something from the value and get the value changed as a prize.

**Comment by niemeyer** [comment]
A few ideas. Very good overall!

File: snap/info.go (hunk 3 above)
Comment (niemeyer) [inline snap/info.go]: Sergio Schvezov, [05.04.16 20:52]
@niemeyer @pedronis during Gramado we killled description for apps; it was originally intended for services (systemd) but it makes little sense as it is not user facing and if it comes to it we can generate one (if the unit name itself doesn't help already)

File: snap/validate.go (hunk 4 above)
Comment (pedronis) [inline snap/validate.go]: ValidateApp makes it easier to refactor minimally the code using verifyAppYaml atm and porting its tests, but yes we should decide wether:

1) always validate
2) validate only when we deal with (not yet installed snap) snap.yaml

File: snappy/binaries.go (hunk 5 above)
Comment (pedronis) [inline snappy/binaries.go]: yes once I have gone over desktop.go and service(s).go

File: snappy/binaries.go (hunk 6 above)
Comment (pedronis) [inline snappy/binaries.go]: agreed, may be for a later branch though

File: snappy/binaries.go (hunk 8 above)
Comment (pedronis) [inline snappy/binaries.go]: that one can be done here

File: snappy/snap_local.go (hunk 12 above)
Comment (pedronis) [inline snappy/snap_local.go]: we can try that but yes would prefer to try a bit later than the shuffling everything, we don't store in names anymore I think, so shouldn't be backward incompatible but may need integration tests work

File: snappy/snap_local.go (hunk 13 above)
Comment (pedronis) [inline snappy/snap_local.go]: @niemeyer it's old code related to sideloading,  the random version used for them was nowhere but the path and then it was used to make the security profile names (so this code captured it), as I wrote in the XXX only some tests need this, I need to see how to change/drop those, anyway when revision is there this will not make sense anyway

File: snap/info.go (hunk 2 above)
Comment (pedronis) [inline snap/info.go]: done

File: snap/info_snap_yaml.go [hunk 15]
```diff
@@ -53,9 +60,23 @@ type slotYaml struct {
...
+
+	Stop        string          `yaml:"stop-command,omitempty"`
+	PostStop    string          `yaml:"post-stop-command,omitempty"`
+	StopTimeout timeout.Timeout `yaml:"stop-timeout,omitempty"`
+
+	RestartCond systemd.RestartCondition `yaml:"restart-condition,omitempty"`
+	SlotNames   []string                 `yaml:"slots,omitempty"`
+	PlugNames   []string                 `yaml:"plugs,omitempty"`
+
+	BusName string `yaml:"bus-name,omitempty"`
```
Comment (zyga) [inline snap/info_snap_yaml.go]: Thinking about this, I wonder what's the user. In the past it was used for frameworks but now that's entirely under the control of interfaces. I have a feeling that we will end up removing this field. At the same time it has some interplay with systemd units. Hmm 

CC @jdstrand

File: snap/info_snap_yaml_test.go [hunk 16]
```diff
@@ -788,3 +789,121 @@ slots:
...
+	y := []byte(`name: fatbinary
+version: 1.0
+`)
+	info, err := snap.InfoFromSnapYaml(y)
+	c.Assert(err, IsNil)
+	c.Assert(info.Architectures, DeepEquals, []string{"all"})
+}
+
+func (s *YamlSuite) TestSnapYamlBadArchitectureParsing(c *C) {
+	data := []byte(`name: fatbinary
```
Comment (zyga) [inline snap/info_snap_yaml_test.go]: Just a nitpick but tests above use ``y`` and this one uses ``data``. Being consistent would look nicer.

File: snap/validate.go (hunk 4 above)
Comment (zyga) [inline snap/validate.go]: I think that having a way not to validate (argument on reading yaml? different function) would be very beneficial. This function is used in various places and it's likely that, as we add better validation, more tests will need boilerplate to just have valid-enough data to perform the operation they are interested in.

Plus, I always consider it good if invalid data can be represented. I'd vote for having two top-levels (parse+validate) or separate, explicit, validate step.

File: snap/validate.go [hunk 17]
```diff
@@ -35,3 +35,50 @@ func ValidateName(name string) error {
...
+}
+
+// appContentWhitelist is the whitelist of legal chars in the "apps"
+// section of snap.yaml
+var appContentWhitelist = regexp.MustCompile(`^[A-Za-z0-9/. _#:-]*$`)
+
+// ValidateApp verifies the content in the app info.
+func ValidateApp(app *AppInfo) error {
+	switch app.Daemon {
+	case "", "simple", "forking", "oneshot", "dbus":
```
Comment (zyga) [inline snap/validate.go]: Nice!

File: snappy/security.go (hunk 1 above)
Comment (zyga) [inline snappy/security.go]: This is obsolete after we switch to interfaces. We don't import anything from ``snappy/`` package in interfaces.

Comment (pedronis) [inline snappy/security.go]: @zyga no, I have the reverse problem, all the functions generating wrappers use this to feed the value into the wrappers themselves, what should I use nowadays? or should we have a method on Info for it instead?

File: snap/info_snap_yaml.go (hunk 15 above)
Comment (jdstrand) [inline snap/info_snap_yaml.go]: This field in 15.04 served two purposes: 1) it was used in the generation of 15.04 simplistic dbus bus policy and 2) if it was specified then 'Type=dbus' would be used in the systemd unit.

'1' is handled completely by interfaces. I think we decided the other day that it is up to the snap developer to choose 'daemon: dbus' (which is supported today) if he/she wants 'Type=dbus' in the systemd unit. As such, I think BusName can and should be removed from the yaml.

Comment (pedronis) [inline snap/info_snap_yaml.go]: ok, something to do after the plumbing is in place, probably not in this branch

File: snappy/snap_local.go (hunk 10 above)
Comment (pedronis) [inline snappy/snap_local.go]: the branch has become quite big already, I'll do it in an immediate subsequent one

File: snap/info_snap_yaml_test.go [hunk 18]
```diff
@@ -788,3 +790,124 @@ slots:
 	c.Check(slot2.Apps, DeepEquals, map[string]*snap.AppInfo{
 		app1.Name: app1, app2.Name: app2})
 }
+
+// type and architectures
+
+func (s *YamlSuite) TestSnapYamlTypeDefault(c *C) {
+	y := []byte(`name: fatbinary
```
Comment (mvo5) [inline snap/info_snap_yaml_test.go]: I feel a bit bad about writing this as this is super nitpick land - this is not a fatbinary, the default here is arch: all which may be a shell script or something else. But given that its a test and is not really used its not really important at all.

Comment (pedronis) [inline snap/info_snap_yaml_test.go]: @mvo5 it's actually a copied test, I fear if we go there I'll never finish

File: snap/validate.go [hunk 19]
```diff
@@ -35,3 +35,47 @@ func ValidateName(name string) error {
...
+	}
+	return nil
+}
+
+// appContentWhitelist is the whitelist of legal chars in the "apps"
+// section of snap.yaml
+var appContentWhitelist = regexp.MustCompile(`^[A-Za-z0-9/. _#:-]*$`)
+
+// ValidateApp verifies the content in the app info.
+func ValidateApp(app *AppInfo) error {
```
Comment (mvo5) [inline snap/validate.go]: I can only second what @zyga said, ValidateApp is really nicely written.

File: snappy/desktop_test.go [hunk 20]
```diff
@@ -124,58 +128,58 @@ Icon=/my/basedir/meep
 }
 
 func (s *SnapTestSuite) TestDesktopFileSanitizeFiltersExec(c *C) {
-	m, err := parseSnapYamlData([]byte(`
```
Comment (mvo5) [inline snappy/desktop_test.go]: Yay for getting rid of `parseSnapYamlData(s, hasConfig)` this bit was always a bit terrible, happy to see it gone.

File: snappy/security.go (hunk 1 above)
Comment (mvo5) [inline snappy/security.go]: Hm, I wonder if getSecurityProfileFromSnapInfo() is slightly better? Longer unfortunately. Even if its just a tempname.

File: snappy/services.go [hunk 21]
```diff
@@ -153,12 +166,11 @@ func addPackageServices(m *snapYaml, baseDir string, inhibitHooks bool, inter in
 		// inhibitHooks mode
 		//
 		// *but* always run enable (which just sets a symlink)
-		serviceName := filepath.Base(generateServiceFileName(m, app))
+		serviceName := filepath.Base(generateServiceFileName(app))
 		sysd := systemd.New(dirs.GlobalRootDir, inter)
-		if !inhibitHooks {
```
Comment (mvo5) [inline snappy/services.go]: If we remove `inhibitHooks` here, we should probably remove the above comment too about inhibitHook mode.

File: snappy/services.go [hunk 22]
```diff
@@ -190,18 +198,22 @@ func addPackageServices(m *snapYaml, baseDir string, inhibitHooks bool, inter in
...
 		}
+		nservices++
 
-		serviceName := filepath.Base(generateServiceFileName(m, app))
+		serviceName := filepath.Base(generateServiceFileName(app))
 		if err := sysd.Disable(serviceName); err != nil {
 			return err
 		}
-		if err := sysd.Stop(serviceName, time.Duration(app.StopTimeout)); err != nil {
+		if err := sysd.Stop(serviceName, serviceStopTimeout(app)); err != nil {
```
Comment (mvo5) [inline snappy/services.go]: Thanks for fixing the serviceStopTimeout()

File: snappy/services.go [hunk 23]
```diff
@@ -212,23 +224,23 @@ func removePackageServices(m *snapYaml, baseDir string, inter interacter) error
...
-		if err := os.Remove(generateBusPolicyFileName(m, app)); err != nil && !os.IsNotExist(err) {
+		if err := os.Remove(generateBusPolicyFileName(app)); err != nil && !os.IsNotExist(err) {
 			logger.Noticef("Failed to remove bus policy file for service %q: %v", serviceName, err)
 		}
 	}
 
 	// only reload if we actually had services
-	// FIXME: filter for services
-	if len(m.Apps) > 0 {
+	if nservices > 0 {
```
Comment (mvo5) [inline snappy/services.go]: Nice!

**Comment by mvo5** [comment]
Looks good, some nitpicking in my comments but really no blockers. Thanks for doing this work!

File: snap/validate.go (hunk 19 above)
Comment (zyga) [inline snap/validate.go]: Why don't we just have ``func (app *AppInfo) Validate() error`` instead? Then we could simply recursively validate everything from snap.Info

**Comment by zyga** [comment]
+1, LGTM
